pkg/plugin/plugin.go

```diff
@@ -44,23 +47,130 @@ type Plugin struct {
+)
+
 // New returns a new *Plugin
 func New(key string, svc kmsiface.KMSAPI, encryptionCtx map[string]string) *Plugin {
+	return newPlugin(
+		key,
+		svc,
+		encryptionCtx,
+		defaultHealthCheckPeriod,
+		defaultErrcBufSize,
+	)
+}
+
+func newPlugin(
+	key string,
+	svc kmsiface.KMSAPI,
+	encryptionCtx map[string]string,
+	checkPeriod time.Duration,
+	errcBuf int,
+) *Plugin {
 	p := &Plugin{
-		svc:   svc,
-		keyID: key,
+		svc:                       svc,
+		keyID:                     key,
+		healthCheckPeriod:         checkPeriod,
+		healthCheckErrc:           make(chan error, errcBuf),
+		healthCheckStopcCloseOnce: new(sync.Once),
+		healthCheckStopc:          make(chan struct{}),
+		healthCheckClosed:         make(chan struct{}),
 	}
 	if len(encryptionCtx) > 0 {
 		p.encryptionCtx = make(map[string]*string)
 	}
 	for k, v := range encryptionCtx {
 		p.encryptionCtx[k] = aws.String(v)
 	}
+	go p.startCheckHealth()
 	return p
 }
 
+func (p *Plugin) startCheckHealth() {
+	zap.L().Info("starting health check routine", zap.String("period", p.healthCheckPeriod.String()))
+	for {
+		select {
+		case <-p.healthCheckStopc:
+			zap.L().Warn("exiting health check routine")
+			p.healthCheckClosed <- struct{}{}
+			return
+		case err := <-p.healthCheckErrc:
+			p.recordErr(err)
+		}
+	}
+}
+
+func (p *Plugin) stopCheckHealth() {
+	p.healthCheckStopcCloseOnce.Do(func() {
+		close(p.healthCheckStopc)
+		<-p.healthCheckClosed
+	})
+}
+
+func (p *Plugin) isRecentlyChecked() (bool, error) {
+	p.lastMu.RLock()
+	err, ts := p.lastErr, p.lastTs
+	never, latest := err == nil && ts.IsZero(), time.Since(ts) < p.healthCheckPeriod
+	p.lastMu.RUnlock()
+	return !never && latest, err
+}
+
+func (p *Plugin) recordErr(err error) {
+	p.lastMu.Lock()
+	p.lastErr, p.lastTs = err, time.Now()
+	p.lastMu.Unlock()
+}
+
+// Health checks KMS API availability.
+//
+// The goal is to:
+//  1. not incur extra KMS API call if plugin "Encrypt" method has already
+//  2. return latest health status (cached KMS status must reflect the current)
+//
+// The error is sent via channel and consumed by goroutine.
+// The error channel may be full and block, when there are too many failures.
+// The error channel may be empty and block, when there's no failure.
+// To handle those two cases, keep track latest health check timestamps.
+//
+// Call KMS "Encrypt" API call iff:
+//  1. there was never a health check done
+//  2. there was no health check done for the last "healthCheckPeriod"
+//     (only use the cached error if the error is from recent API call)
+//
+func (p *Plugin) Health() error {
+	recent, err := p.isRecentlyChecked()
+	if !recent {
+		_, err = p.Encrypt(context.Background(), &pb.EncryptRequest{Plain: []byte("foo")})
+		p.recordErr(err)
+		if err != nil {
+			zap.L().Warn("health check failed", zap.Error(err))
+		}
+		return err
+	}
+	if err != nil {
+		zap.L().Warn("health check fail", zap.Error(err))
+	} else {
+		zap.L().Info("health check success")
```

Comment:
Should this be debug level? If called every 30 sec and its a healthy operation, it seems a little excessive